src/unit/test_fifo.c

```diff
@@ -0,0 +1,248 @@
+                    for (int i = 0; i < qPop; i++) {
+                        void *ptr;
+                        TEST_EXPECT(fifoPop(q, &ptr) && PTR_TO_INT(ptr) == popQValue++);
+                    }
+
+                    intptr_t popQ2Value = pushValue;
+                    for (int i = 0; i < q2Len; i++) fifoPush(q2, INT_TO_PTR(pushValue++));
+                    for (int i = 0; i < q2Pop; i++) {
+                        void *ptr;
+                        TEST_EXPECT(fifoPop(q2, &ptr) && PTR_TO_INT(ptr) == popQ2Value++);
+                    }
+
+                    fifoJoin(q, q2);
+                    TEST_EXPECT(fifoLength(q) == (qLen - qPop) + (q2Len - q2Pop));
+                    TEST_EXPECT(fifoLength(q2) == 0);
+
+                    fifo *temp = fifoPopAll(q); /* Exercise fifoPopAll also */
+                    TEST_EXPECT(fifoLength(temp) == (qLen - qPop) + (q2Len - q2Pop));
+                    TEST_EXPECT(fifoLength(q) == 0);
+
+                    for (int i = 0; i < (qLen - qPop); i++) {
+                        void *ptr;
+                        TEST_EXPECT(fifoPop(temp, &ptr) && PTR_TO_INT(ptr) == popQValue++);
+                    }
+                    for (int i = 0; i < (q2Len - q2Pop); i++) {
+                        void *ptr;
+                        TEST_EXPECT(fifoPop(temp, &ptr) && PTR_TO_INT(ptr) == popQ2Value++);
+                    }
+                    TEST_EXPECT(fifoLength(temp) == 0);
+
+                    fifoDelete(temp);
+                }
+            }
+        }
+    }
+
+    fifoDelete(q);
+    fifoDelete(q2);
+    return 0;
+}
+
+/* Set to 1 to prove that FIFO outperforms ADLIST. This test will exercise both.
+ * The test will (intentionally) fail, printing the results as failed assertions. */
+#define COMPARE_PERFORMANCE_TO_ADLIST 0
+
+#include "../adlist.h"
+#include "../monotonic.h"
+
+const int LIST_ITEMS = 10000;
+
+static void exerciseList(void) {
+    list *q = listCreate();
+    for (intptr_t i = 0; i < LIST_ITEMS; i++) {
+        listAddNodeTail(q, INT_TO_PTR(i));
+    }
+    TEST_EXPECT(listLength(q) == (unsigned)LIST_ITEMS);
+    for (intptr_t i = 0; i < LIST_ITEMS; i++) {
+        listNode *node = listFirst(q);
+        listDelNode(q, node);
+        TEST_EXPECT(listNodeValue(node) == INT_TO_PTR(i));
+    }
+    TEST_EXPECT(listLength(q) == 0u);
+    listRelease(q);
+}
+
+static void exerciseFifo(void) {
+    fifo *q = fifoCreate();
+    for (intptr_t i = 0; i < LIST_ITEMS; i++) {
+        fifoPush(q, INT_TO_PTR(i));
+    }
+    TEST_EXPECT(fifoLength(q) == LIST_ITEMS);
+    for (intptr_t i = 0; i < LIST_ITEMS; i++) {
+        void *ptr;
+        TEST_EXPECT(fifoPop(q, &ptr) && ptr == INT_TO_PTR(i));
+    }
+    TEST_EXPECT(fifoLength(q) == 0);
+    fifoDelete(q);
+}
+
+int test_fifoComparePerformance(int argc, char *argv[], int flags) {
+    UNUSED(argc);
+    UNUSED(argv);
+    UNUSED(flags);
+
+    if (COMPARE_PERFORMANCE_TO_ADLIST) {
```

Comment:
```suggestion
    if (argc > 2 && !strcasecmp(argv[1], "--compare-performance-to-adlist") { 
```
argv is supposed to be used to optionally execute tests.

Reply:
Done. To run the performance comparison test, use `./valkey-unit-tests --single test_fifo.c --compare-performance-to-adlist`